group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorShard.java

```diff
@@ -47,4 +47,15 @@ default void onNewMetadataImage(MetadataImage newImage, MetadataDelta delta) {}
      * any post unloading operations.
      */
     default void onUnloaded() {}
+
+    /**
+     * Replay a record to update the state machine.
+     *
+     * @param record The record to replay.
+     */
+    default void replay(
+        long producerId,
+        short producerEpoch,
+        U record
+    ) throws RuntimeException {}
```

Comment:
I still don't get why we need a default implementation here. I think that it should be:
```
void replay(
        long producerId,
        short producerEpoch,
        U record
) throws RuntimeException;
```

Reply:
ah, i misunderstood. thanks for pointing it out